src/vitamin.cpp

```diff
@@ -0,0 +1,74 @@
+#include "vitamin.h"
+
+#include <map>
+
+#include "debug.h"
+
+static std::map<vitamin_id, vitamin> vitamins_all;
+
+template<>
+bool string_id<vitamin>::is_valid() const
+{
+    return vitamins_all.count( *this );
+}
+
+template<>
+const vitamin &string_id<vitamin>::obj() const
+{
+    const auto found = vitamins_all.find( *this );
+    if( found == vitamins_all.end() ) {
+        debugmsg( "Tried to get invalid vitamin: %s", c_str() );
+        static const vitamin null_vitamin{};
+        return null_vitamin;
+    }
+    return found->second;
+}
+
+efftype_id vitamin::effect( int level ) const
+{
+    for( const auto &e : deficiency_ ) {
+        if( level <= e.second ) {
+            return e.first;
+        }
+    }
+    for( const auto &e : excess_ ) {
+        if( level >= e.second ) {
+            return e.first;
+        }
+    }
+    return NULL_ID;
+}
+
+void vitamin::load_vitamin( JsonObject &jo )
+{
+    vitamin vit;
+
+    vit.id_ = vitamin_id( jo.get_string( "id" ) );
+    vit.name_ = jo.get_string( "name" );
+    vit.min_ = jo.get_int( "min" );
+    vit.max_ = jo.get_int( "max", 0 );
+
+    auto def = jo.get_array( "deficiency" );
+    while( def.has_more() ) {
+        auto e = def.next_array();
+        vit.deficiency_.emplace_back( efftype_id( e.get_string( 0 ) ), e.get_int( 1 ) );
+    }
+    auto exc = jo.get_array( "excess" );
+    while( exc.has_more() ) {
+        auto e = exc.next_array();
+        vit.excess_.emplace_back( efftype_id( e.get_string( 0 ) ), e.get_int( 1 ) );
+    }
+
+    vitamins_all[ vit.id_ ] = vit;
```

Comment:
Say, a mod contains a vitamin definition. In this case all fields will be reset with no warning, no value will be kept.